pkg/chaosdaemon/iptables_server.go

```diff
@@ -42,144 +42,164 @@ func (s *daemonServer) SetIptablesChains(ctx context.Context, req *pb.IptablesCh
+				}, rule)
+				if err != nil {
+					log.Error(err, "error while add iptables chains")
+					return err
+				}
+			}
+		} else {
+			err := iptables.ensureRule(&iptablesChain{
+				Name: chain.ChainName,
+			}, rule)
+			if err != nil {
+				log.Error(err, "error while add iptables chains")
+				return err
+			}
+		}
+	case pb.Chain_DELETE:
+		log.Info("Delete Chain ", chain.ChainName)
+		err := iptables.flushIptablesChain(&iptablesChain{
+			Name: chain.ChainName,
+			Rule: "",
+		})
 		if err != nil {
+			log.Error(err, "error while add iptables chains")
 			return err
 		}
+		err = iptables.ensureRule(&iptablesChain{
+			Name: chain.ChainName,
+		}, fmt.Sprintf("-X %s ", chain.ChainName))
+	default:
+		return fmt.Errorf("error no command in iptables chains")
 	}
-
 	return nil
 }
 
-func (iptables *iptablesClient) setIptablesChain(chain *pb.Chain) error {
-	var matchPart string
-	if chain.Direction == pb.Chain_INPUT {
-		matchPart = "src"
-	} else if chain.Direction == pb.Chain_OUTPUT {
-		matchPart = "dst"
+func parseAddChain(chain *pb.Chain) (string, error) {
+	var rule string
+	if chain.Table != "" {
+		rule += fmt.Sprintf("-t %s ", chain.Table)
+	}
+	if chain.ChainName != "" {
+		rule += fmt.Sprintf("-A %s -p tcp ", chain.ChainName)
 	} else {
-		return fmt.Errorf("unknown chain direction %d", chain.Direction)
+		return "", fmt.Errorf("add chain but no chain name")
 	}
-
-	rules := []string{}
-	for _, ipset := range chain.Ipsets {
-		rules = append(rules, fmt.Sprintf("-A %s -m set --match-set %s %s -j DROP -w 5", chain.Name, ipset, matchPart))
+	if chain.SourceAddress != "" {
+		rule += fmt.Sprintf("-s %s ", chain.SourceAddress)
 	}
-	err := iptables.createNewChain(&iptablesChain{
-		Name:  chain.Name,
-		Rules: rules,
-	})
-	if err != nil {
-		return err
+	if chain.Sport != "" {
+		rule += fmt.Sprintf("--sport %s ", chain.Sport)
 	}
-
-	if chain.Direction == pb.Chain_INPUT {
-		err := iptables.ensureRule(&iptablesChain{
-			Name: "CHAOS-INPUT",
-		}, "-A CHAOS-INPUT -j "+chain.Name)
-		if err != nil {
-			return err
+	if chain.DestinationAddress != "" {
+		rule += fmt.Sprintf("-d %s ", chain.DestinationAddress)
+	}
+	if chain.Dport != "" {
+		rule += fmt.Sprintf("--dport %s ", chain.Dport)
+	}
+	if chain.ToPorts != "" {
+		rule += fmt.Sprintf("--to-ports %s ", chain.ToPorts)
+	}
+	if chain.Probability != "" {
+		rule += fmt.Sprintf("--probability %s ", chain.Dport)
+	}
+	if chain.IpsetsName != "" {
+		var matchPart string
+		if chain.Direction == pb.Chain_INPUT {
+			matchPart = "src"
+		} else if chain.Direction == pb.Chain_OUTPUT {
+			matchPart = "dst"
+		} else {
+			return "", fmt.Errorf("unknown chain direction %d", chain.Direction)
 		}
-	} else if chain.Direction == pb.Chain_OUTPUT {
-		iptables.ensureRule(&iptablesChain{
-			Name: "CHAOS-OUTPUT",
-		}, "-A CHAOS-OUTPUT -j "+chain.Name)
-		if err != nil {
-			return err
+		if len(chain.Ipsets) > 1 {
+			rule += fmt.Sprintf("-m set --match-set %s %%s -w 5 ", matchPart)
```

Comment:
seems lack some argument

Reply:
The second placeholder was "%%s" that was just a real percent symbol i use it in the [https://github.com/chaos-mesh/chaos-mesh/pull/938/files/bf54990108dc99f8b4f1a78a4c1fe9cd9439100d#diff-7f7fb0cc6d572e887d964b8933b22145R82]

Reply:
ok, how about adding a comment

Reply:
yes, I agree with you.